### DISCUSSION TITLE
NotEnoughMemoryException when sending a large size message [14820]

### DISCUSSION THREAD

**Discussion by Nir-Az** [eProsima/Fast-DDS#2707] [Q&A]
Hi,

I am trying to send an image throw a FastDDS writer and I get this error:

```
RTPSWriter:Serialization returns false
```

When I debug it I see that FastDDS throws `NotEnoughMemoryException` exception
My IDL contains `sequence<octet> raw_data;` and I am filling it with a ~2.5 [MB] data

I guess it's got to do with the `MemoryManagementPolicy` but it is not inforced.
If the message is unbounded, how can FastDDS accept `PREALLOCATED_MEMORY_MODE`?

Can someone please explain better the difference between `PREALLOCATED_WITH_REALLOC_MEMORY_MODE` and `DYNAMIC_REUSABLE_MEMORY_MODE`?

What is more efficient for streaming?

My use case is supporting streaming of different size images (user control) but once the size is chosen , it will send the same message size every time.

Thanks,
Nir

**Comment by JLBuenoLopez** [comment]
Hi @Nir-Az,

It is responsibility of the user to configure correctly its Fast DDS application. Even if the message is unbounded, Fast DDS allows using `PREALLOCATED_MEMORY_MODE` because internally Fast DDS allocates a estimated sample size. If at runtime the sample size is bigger that the estimated one, then the exception you mention is thrown.

Regarding the difference between `PREALLOCATED_WITH_REALLOC_MEMORY_MODE` and `DYNAMIC_REUSABLE_MEMORY_MODE`, I encourage you to read [Fast DDS MemoryManagementPolicy documentation section](https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/core/policy/eprosimaExtensions.html#memorymanagementpolicy). The main difference is that `PREALLOCATED_WITH_REALLOC_MEMORY_MODE` preallocates a certain sample size and if that size is not enough, reallocation is allowed. `DYNAMIC_REUSABLE_MEMORY_MODE` does not preallocate any memory when Fast DDS is initialized. All memory needed is allocated dynamically. Also, when the memory is released, instead of releasing it to the system, it is saved in a pool so it can be reused. This is the main difference with `DYNAMIC_RESERVE_MEMORY_MODE` which does not reuse the released memory. `DYNAMIC_REUSABLE_MEMORY_MODE` memory consumption may be lower than the `PREALLOCATED_WITH_REALLOC_MEMORY_MODE` one depending on other QoS configuration such as [HistoryQosPolicy](https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/core/policy/standardQosPolicies.html#historyqospolicy), [ResourceLimitsQosPolicy](https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/core/policy/standardQosPolicies.html#resourcelimitsqospolicy), etc.

I am afraid I cannot help you with your last question. You should probably test your application using both modes and see what works best for you. Probably, in a steady state, it would not matter.
